improve compiler error message for failed overload resolution

smoothdeveloper opened the pull request:
Work related to method overload resolution error messages: #6578.

* [x] implementing changes to match the original description
   * [x] find other error message cases
* [x] add sufficient test coverage in the new test suite
   * [x] overloads with out / ref / byref parameters (covered by some fsharpqa migrated tests)
   * [x] check the impact of generic in error display  (covered by some fsharpqa migrated + new tests)
   * [x] no redundant spacing among the sections of the message (checked in situ and enforced in .bsl files)
  * [x] add more cases, involving same technique as FSharpPlus (or else) and making sure that relevant type mistmatch information is not lost
* [ ] consider listing the overloads in case of argument count mismatch (right now there is some logic to propose a correct number of parameters) -> separate PR
* [x] review of PR -> MSFT
* [x] finalize wording / name of errors in the code
* [x] integrate review feedback
* [x] investigate the documentation / guides for parts that could be impacted by error message changes -> MSFT

smoothdeveloper commented on tests/fsharpqa/Source/ErrorMessages/OverloadResolution/E_System.Convert.ToString.OverloadList.fs:
```diff
@@ -1,4 +1,3 @@
 System.Convert.ToString('a',0) |> ignore
 
 //<Expects id="FS0039" status="error" span="(1,1-1,1)">is not compatible with type</Expects>
```
this test is still red (supposed to be compatible with current message for now), I'll need to get Run.pl to work on my machine to work with this (help on #6593 appreciated).

Can I see the output of that particular test from the CI beyond `-- failed` ? https://dev.azure.com/dnceng/public/_build/results?buildId=163469

smoothdeveloper commented:
Few notes for myself & reviewers that would like to validate my approach.

I've been stepping around the overload resolution code, for now I'm considering this option for implementation:

* Change `ConstraintSolver.ResolveOverloading` `callerArgCounts` argument into `callerArgs` at this call site

https://github.com/Microsoft/visualfsharp/blob/c4c1d9cf1097955a71c7cbc328a6b5c672a34bda/src/fsharp/TypeChecker.fs#L9889-L9900

(need to check the other call site to `ResolveOverloading` happening in `SolveMemberConstraint` how change this argument can be handled)

* using this information, assemble the top of the message, which will contain the method application given parameter types in `ResolveOverloading`'s `failOverloading` local function.

* continue assembling the rest of the message where it is currently handled in `CompileOps.OutputPhasedErrorR` appending the overload list in a compact form

I was considering building the whole message in `CompileOps.OutputPhasedErrorR` but the change can be implemented without touching this aspect and make review easier.

smoothdeveloper commented:
I'm trying to understand the reason for `callerArgs` type to be `(CallerArg<Expr> list * CallerNamedArg<Expr> list) list`: a list of tuple of lists.

https://github.com/Microsoft/visualfsharp/blob/c4c1d9cf1097955a71c7cbc328a6b5c672a34bda/src/fsharp/TypeChecker.fs#L9879

I'd see the caller arguments as either:
* `(CallerArg | CallerNamedArg) list`
* `(CallerArg list * CallerNamedArg list)`

I'm not sure what is the extra nesting in a list is used for?

smoothdeveloper commented:
Just a note listing the fsharpqa tests that started failing when I changed the error message a bit (https://dev.azure.com/dnceng/public/_build/results?buildId=164032).

I want to investigate how those flow through `ConstraintSolver.ResolveOverloading`.

> Conformance\Expressions\SyntacticSugar (E_Slices01.fs.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_OneTypeVariable03.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_OneTypeVariable03rec.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_TwoDifferentTypeVariables01.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_TwoDifferentTypeVariables01rec.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_TwoDifferentTypeVariablesGen00rec.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_TwoEqualTypeVariables02.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_TwoEqualYypeVariables02rec.fs) -- failed
Conformance\InferenceProcedures\TypeInference (E_LeftToRightOverloadResolution01.fs) -- failed
Conformance\InferenceProcedures\WellFormednessChecking (E_Clashing_Values_in_AbstractClass01.fs) -- failed
Conformance\InferenceProcedures\WellFormednessChecking (E_Clashing_Values_in_AbstractClass03.fs) -- failed
Conformance\InferenceProcedures\WellFormednessChecking (E_Clashing_Values_in_AbstractClass04.fs) -- failed
ErrorMessages\OverloadResolution (E_System.Convert.ToString.OverloadList) -- failed

smoothdeveloper commented:
open/close principle for CI

smoothdeveloper commented:
the work is progressing, there are new tests failling in fsharpqa, that i'll probably move in this PR

> Conformance\Expressions\Type-relatedExpressions (E_RigidTypeAnnotation03.fsx) -- failed
> Conformance\InferenceProcedures\TypeInference (E_TwoDifferentTypeVariablesGen00.fs) -- failed
> Conformance\LexicalAnalysis\SymbolicOperators (E_LessThanDotOpenParen001.fs) -- failed


![image](https://user-images.githubusercontent.com/87944/57002764-8c8eea00-6b76-11e9-8ae1-5055e2b43b59.png)

This is going through a code path I haven't yet stepped through (units of measure show in the test) and need adjustment to the message building logic, I'll look further toward the end of the week.

cartermp commented:
@smoothdeveloper Wonderful!

smoothdeveloper commented:
VS error box (left=after,right=before):
![vs error box](https://user-images.githubusercontent.com/87944/57086914-e21ce100-6cb3-11e9-8285-63e220910938.png)

VS Tooltip before:
![vs tooltip before](https://user-images.githubusercontent.com/87944/57086916-e21ce100-6cb3-11e9-97a4-8f9c38c80b92.png)
VS Tooltip after:
![vs tooltip after](https://user-images.githubusercontent.com/87944/57086915-e21ce100-6cb3-11e9-9241-8b33a3c88fcd.png)

VS Tooltip (alternate message, filtered on ref types) before (single line message in tooltip woes):
![vs tooltip reftype before](https://user-images.githubusercontent.com/87944/57086918-e2b57780-6cb3-11e9-9e4b-2957714c6640.png)

VS Tooltip (alternate message, filtered on ref types) after:
![vs tooltip reftype after](https://user-images.githubusercontent.com/87944/57086917-e21ce100-6cb3-11e9-9fd7-b27c70e3291e.png)

The experience in FSI is also good.

I've fixed the windows desktop failing tests but now new fsharpqa test is failing. Pushing some minor changes to do a full run again.

cartermp commented:
@smoothdeveloper Absolutely wonderful!

smoothdeveloper commented:
This is green but I had to disable the fsharpqa optimization / inlining test: https://github.com/Microsoft/visualfsharp/pull/6596/commits/8514f092a2d469707bdf9dc18a843d3c2e743412

smoothdeveloper commented:
Green again and ready for review.

cartermp commented:
@KevinRansom can you take a look?

dsyme commented on tests/fsharp/typecheck/sigs/neg61.bsl:
```diff
@@ -83,17 +83,18 @@ neg61.fs(156,21,156,22): typecheck error FS3147: This 'let' definition may not b
-    'System.Collections.Generic.IEnumerable<'a>'    
-.
+neg61.fs(174,22,174,23): typecheck error FS0041: No overloads match for method 'Source'.
+
```
Why so many blank lines?

smoothdeveloper replied:
I mention the issue in this comment: https://github.com/Microsoft/visualfsharp/pull/6654#issuecomment-488056692

the standard output / error redirect used in the test lead to this kind of output, although when using FSI in VS or command line, or in VS editing workflow, the messages do not have those extra CR.

dsyme commented on tests/fsharp/typecheck/sigs/neg20.bsl:
```diff
@@ -256,17 +264,21 @@ neg20.fs(184,34,184,39): typecheck error FS0001: This expression was expected to
+Arguments given:
+ - string
+
+ - obj
```
I find this a little odd - Could we format as
```
Arguments given: string * obj
```
or
```
Known types of arguments: string * obj
```

smoothdeveloper replied:
@dsyme, what formatting should be used for named arguments at call site?

Should we bother distinguishing if the method is defined with untupled arguments versus tupled?

dsyme commented on src/fsharp/MethodCalls.fs:
```diff
@@ -99,6 +102,21 @@ type CallerNamedArg<'T> =
+type CallerArgs<'T> = 
+    { Unnamed: CallerArg<'T> list list
+      Named: CallerNamedArg<'T> list list }
+with
```
Remove this `with` please, we never use that for augmentations in the compiler

dsyme commented on src/fsharp/MethodCalls.fs:
```diff
@@ -43,13 +43,16 @@ open FSharp.Compiler.ExtensionTyping
+      IsOptional: bool 
+      Expr: 'T }
     
+module CallerArg =
```
Please remove these functions or change to static `Create` members, we basically never use this style of "extra module for constructors of a record type" in the compiler

dsyme commented on src/fsharp/CompileOps.fs:
```diff
@@ -755,23 +750,52 @@ let OutputPhasedErrorR (os: StringBuilder) (err: PhasedDiagnostic) (suggestNames
+      | UnresolvedOverloading(denv, callerArgs, failure, m) ->
+          let nl = System.Environment.NewLine
+          let displayArgType (name, ttype) =
+              let typeDisplay = NicePrint.prettyStringOfTy denv ttype
```
The technique to display types needs to be adjusted.  The rule is "format all the types in one go".  This ensures that type inference variables that are actually the same across the different types get formatted to the same text.  See `minimalStringsOfTwoTypes` for example for how we format two types (which in that case are then going to be compared).  There are also helpers to format multiple types I believe.

For test cases, you should I think notice the difference when types cross arguments are linked, e.g. `C.M(x,x)` where each `x` has an unknown inference type, or `C.M(x,x,y,y)`

smoothdeveloper replied:
> See `minimalStringsOfTwoTypes` for example for how we format two types (which in that case are then going to be compared). There are also helpers to format multiple types I believe.

I'm going to look further into this, I've just hit https://github.com/microsoft/visualfsharp/issues/6725.

> For test cases, you should I think notice the difference when types cross arguments are linked, e.g. C.M(x,x) where each x has an unknown inference type, or C.M(x,x,y,y)

Could you tell if:

```fsharp
type C() =
    static member M(a, b) = a + b
    static member M(a, b, c) = a + b + c
    static member M(a, b, c, d, e) = float (a + b + c + d) * e
    static member M(a, b, c, d, e) = decimal (a + b + c + d) * e
    static member M(a, b, c, d, e) = float32 (a + b + c + d) * e
C.M(1, 2, 3, 4, 5)
```

is a relevant case for what you described?

it currently gives:

![image](https://user-images.githubusercontent.com/87944/57675308-73565680-75d6-11e9-8f91-cd5b0c76a0c3.png)

and current state of this PR:

![image](https://user-images.githubusercontent.com/87944/57675604-4b1b2780-75d7-11e9-90a7-a4f9f4e00afc.png)

When generics are involved, I'm not sure the changes make it more readable, do we want to do something special for those cases?

smoothdeveloper replied:
@dsyme:

> The technique to display types needs to be adjusted. The rule is "format all the types in one go". This ensures that type inference variables that are actually the same across the different types get formatted to the same text.

Looking at original code, I don't see that rule being considered, also, does it make sense to consolidate generic arguments names (`'a`, `'b`, etc.)  over a list of many overloads in the error message?

It seems to me it would lead to messages with generic arguments named `'z` without having 26 type parameters on a given overload.

Please see in original code, the stringification happens in a List.map and doesn't seem to pass any state that would be mutated (or I've overlooked that):

https://github.com/dotnet/fsharp/pull/6596/files?file-filters%5B%5D=.fs&file-filters%5B%5D=.fsi&file-filters%5B%5D=.fsx#diff-641da7c52ffc888699888ebd03559cb5L175

https://github.com/dotnet/fsharp/pull/6596/files?file-filters%5B%5D=.fs&file-filters%5B%5D=.fsi&file-filters%5B%5D=.fsx#diff-641da7c52ffc888699888ebd03559cb5L2552

So in the context of type id "unification/prettification", I don't see this PR making any change of behaviour compared to existing code, am I still overlooking something on that matter?

dsyme replied:
Can you prettify the group of arguments for a single overload in a single prettification? (Not all the overloads at once)

smoothdeveloper replied:
@dsyme, right makes more sense than what I said wouldn't work :slightly_smiling_face:.

AFAIU, this is handled by call to `stringOfMethInfo`, call which has remained in this PR.

It ends up there: 
https://github.com/dotnet/fsharp/blob/ccb913d3a05863e5b1861d64994ffb97ed498855/src/fsharp/NicePrint.fs#L1381-L1397

smoothdeveloper replied:
I found a case where the "Known types of arguments" would hit the issue @dsyme is concerned about:

```fsharp

type A =
  static member Foo(a: 'a, b: 'a -> 'b, c: 'a -> 'b, d: 'a -> 'b) : 'b = b a
  static member Foo(a: 'a, b: 'a -> 'b, c: 'b -> 'c, d: 'c -> 'd) : 'd = d (c( b a))

let inline f a b c =
  A.Foo(1., a, b, c)
```

printing

>  A unique overload for method 'Foo' could not be determined based on type information prior to this program point. A type annotation may be needed.
>
> Known types of arguments: __float * 'a * 'a * 'a__
>
>Candidates:
> - static member A.Foo : a:'a * b:('a -> 'b) * c:('a -> 'b) * d:('a -> 'b) -> 'b
> - static member A.Foo : a:'a * b:('a -> 'b) * c:('b -> 'c) * d:('c -> 'd) -> 'd

I gather we should have __float * 'a * 'b * 'c__ here.

Going to look more at NicePrint.fs and `Prettify*` functions in tastops.fs.

dsyme commented on src/fsharp/ConstraintSolver.fs:
```diff
@@ -155,25 +156,36 @@ type ContextInfo =
+  amap : ImportMap
+  error: exn
+}
+with
```
Remove this `with` and format the record like we do all the others, 
```fsharp
[<Struct>]
type OverloadInformation = 
    {
      methodSlot: CalledMeth<Expr>
      amap : ImportMap
      error: exn
    }
    member ...
```

dsyme commented on src/fsharp/ConstraintSolver.fs:
```diff
@@ -2536,27 +2562,26 @@ and ResolveOverloading
-                    None, ErrorD (failOverloading msg []), NoTrace
+                    let methods = 
+                        let getMethodSlotsAndErrors =
+                          function | methodSlot, []      -> List.singleton {methodSlot = methodSlot; error = Unchecked.defaultof<exn>; amap = amap}
```
Please use `match` not `function`

dsyme replied:
Nit - Anything involving `List.map` and `List.singleton` is almost always better as a list expression, e.g. this looks equivalent to:
```
let getMethodSlotsAndErrors (methodSlot, errors) =
    [ match errors with 
      | [] -> yield {methodSlot = methodSlot; error = Unchecked.defaultof<exn>; amap = amap}
      | errors  -> for error in errors do yield {methodSlot = methodSlot; error = error; amap = amap} ]
```

gusty commented:
@smoothdeveloper this is great but am I getting it wrong or your PR will remove the "the type A is not compatible with type B" part of the message?

Because I find that part very useful when debugging overload errors.

smoothdeveloper commented:
@gusty yes it removes the "type subsumption" errors from the error message in context of overloaded methods, it doesn't impact other places also using that message.

Do you have specific cases in mind of error messages that become more obscure without that part? I'd be happy to run those with this PR to gather further feedback.

Maybe those type subsumption errors could be kept at the end?

My concern is that interleaved, they were the cause of making those messages oddly laid out in order for the user to have overview of relevant overloads.

gusty commented:
I don't have cases in mind right now, but for sure I had them and that part of the message helped me to focus on which arguments are not matching.

Normally when all overloads are "at the same level" it doesn't matter, but sometimes you have an overload that cover more generic cases than the other, or that is not matching because another parameter.

If I ever see a case I will post it, in the meantime let's think how can we re-organize the error message in such a way that that part is not completely removed.

smoothdeveloper commented:
@gusty 

with this code, typo intended in get_Zero:

```fsharp
type MonoidSample = 
    | MonoidSample of int

    static member getr_Zero () = MonoidSample 0
    //static member (+) (MonoidSample(x), MonoidSample(y)) = MonoidSample (x+y)
    static member (+) (MonoidSample(x), y) = MonoidSample (x+y), 1
    static member (+) (x,MonoidSample(y)) = MonoidSample (x+y)
    static member (+) (x, y) = MonoidSample (x+y), 3

let a = MonoidSample 1
let b = MonoidSample 2
let c : MonoidSample = zero
```

this PR:

```
C:\home\tmp\test.fsx(45,24): error FS0001: No overloads match for method 'Zero'.

Available overloads:
 - static member Control.Zero.Zero :  ^t * Internals.Default1 ->  ^t when  ^t : (static member get_Zero : ->  ^t)
 - static member Control.Zero.Zero :  ^t * Internals.Default1 -> ('a1 -> 'a1) when  ^t : null and  ^t : struct
 - static member Control.Zero.Zero :  ^t * Internals.Default2 ->  ^t when (Control.FromInt32 or  ^t) : (static member FromInt32 :  ^t * Control.FromInt32 -> int32 ->  ^t)
 - static member Control.Zero.Zero :  ^t * Internals.Default2 -> ('a1 -> 'a1) when  ^t : null and  ^t : struct
 - static member Control.Zero.Zero :  ^t * Internals.Default3 ->  ^t when  ^t : (static member get_Empty : ->  ^t)
 - static member Control.Zero.Zero : 'a array * Control.Zero -> 'a array
 - static member Control.Zero.Zero : 'a list * Control.Zero -> 'a list
 - static member Control.Zero.Zero : 'a option * Control.Zero -> 'a option
 - static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d *  ^e *  ^f *  ^g) * Control.Zero ->  ^a *  ^b *  ^c *  ^d *  ^e *  ^f *  ^g when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d) and (Control.Zero or  ^e) : (static member Zero :  ^e * Control.Zero ->  ^e) and (Control.Zero or  ^f) : (static member Zero :  ^f * Control.Zero ->  ^f) and (Control.Zero or  ^g) : (static member Zero :  ^g * Control.Zero ->  ^g)
 - static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d *  ^e *  ^f) * Control.Zero ->  ^a *  ^b *  ^c *  ^d *  ^e *  ^f when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d) and (Control.Zero or  ^e) : (static member Zero :  ^e * Control.Zero ->  ^e) and (Control.Zero or  ^f) : (static member Zero :  ^f * Control.Zero ->  ^f)
 - static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d *  ^e) * Control.Zero ->  ^a *  ^b *  ^c *  ^d *  ^e when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d) and (Control.Zero or  ^e) : (static member Zero :  ^e * Control.Zero ->  ^e)
 - static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d) * Control.Zero ->  ^a *  ^b *  ^c *  ^d when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d)
 - static member Control.Zero.Zero : ( ^a *  ^b *  ^c) * Control.Zero ->  ^a *  ^b *  ^c when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c)
 - static member Control.Zero.Zero : ( ^a *  ^b) * Control.Zero ->  ^a *  ^b when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b)
 - static member Control.Zero.Zero : ('T ->  ^Monoid) * Control.Zero -> ('T ->  ^Monoid) when (Control.Zero or  ^Monoid) : (static member Zero :  ^Monoid * Control.Zero ->  ^Monoid)
 - static member Control.Zero.Zero : Async< ^a> * Control.Zero -> Async< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)
 - static member Control.Zero.Zero : Internals.DmStruct * Control.Zero -> Internals.DmStruct
 - static member Control.Zero.Zero : Internals.Id<'a> * Control.Zero -> Internals.Id< ^a1> when (Control.Zero or  ^a1) : (static member Zero :  ^a1 * Control.Zero ->  ^a1)
 - static member Control.Zero.Zero : Map<'a,'b> * Control.Zero -> Map<'a,'b> when 'a : comparison
 - static member Control.Zero.Zero : Quotations.Expr< ^a> * Control.Zero -> Quotations.Expr< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)
 - static member Control.Zero.Zero : ResizeArray<'a> * Control.Zero -> ResizeArray<'a>
 - static member Control.Zero.Zero : Set<'a> * Control.Zero -> Set<'a> when 'a : comparison
 - static member Control.Zero.Zero : System.Collections.Generic.Dictionary<'a,'b> * Control.Zero -> System.Collections.Generic.Dictionary<'a,'b> when 'a : equality
 - static member Control.Zero.Zero : System.Collections.Generic.IDictionary<'a,'b> * Control.Zero -> System.Collections.Generic.IDictionary<'a,'b> when 'a : equality
 - static member Control.Zero.Zero : System.Collections.Generic.IEnumerator<'a> * Control.Zero -> System.Collections.Generic.IEnumerator<'a>
 - static member Control.Zero.Zero : System.Collections.Generic.IReadOnlyDictionary<'a,'b> * Control.Zero -> System.Collections.Generic.IReadOnlyDictionary<'a,'b> when 'a : equality
 - static member Control.Zero.Zero : System.Lazy< ^a> * Control.Zero -> Lazy< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)
 - static member Control.Zero.Zero : System.Text.StringBuilder * Control.Zero -> System.Text.StringBuilder
 - static member Control.Zero.Zero : System.Threading.Tasks.Task< ^a> * Control.Zero -> System.Threading.Tasks.Task< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)
 - static member Control.Zero.Zero : System.TimeSpan * Control.Zero -> System.TimeSpan
 - static member Control.Zero.Zero : System.Tuple< ^a> * Control.Zero -> System.Tuple< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)
 - static member Control.Zero.Zero : seq<'a> * Control.Zero -> seq<'a>
 - static member Control.Zero.Zero : string * Control.Zero -> string
 - static member Control.Zero.Zero : t: ^t * Control.Zero ->  ^t when  ^t : (member get_Item1 :  ^t ->  ^t1) and  ^t : (member get_Item2 :  ^t ->  ^t2) and  ^t : (member get_Item3 :  ^t ->  ^t3) and  ^t : (member get_Item4 :  ^t ->  ^t4) and  ^t : (member get_Item5 :  ^t ->  ^t5) and  ^t : (member get_Item6 :  ^t ->  ^t6) and  ^t : (member get_Item7 :  ^t ->  ^t7) and  ^t : (member get_Rest :  ^t ->  ^tr) and (Control.Zero or  ^t1) : (static member Zero :  ^t1 * Control.Zero ->  ^t1) and (Control.Zero or  ^t2) : (static member Zero :  ^t2 * Control.Zero ->  ^t2) and (Control.Zero or  ^t3) : (static member Zero :  ^t3 * Control.Zero ->  ^t3) and (Control.Zero or  ^t4) : (static member Zero :  ^t4 * Control.Zero ->  ^t4) and (Control.Zero or  ^t5) : (static member Zero :  ^t5 * Control.Zero ->  ^t5) and (Control.Zero or  ^t6) : (static member Zero :  ^t6 * Control.Zero ->  ^t6) and (Control.Zero or  ^t7) : (static member Zero :  ^t7 * Control.Zero ->  ^t7) and (Control.Zero or  ^tr) : (static member Zero :  ^tr * Control.Zero ->  ^tr)
 - static member Control.Zero.Zero : unit * Control.Zero -> unit
```

existing compiler:
```
C:\home\tmp\test.fsx(45,24): error FS0001: No overloads match for method 'Zero'. The available overloads are shown below.
Possible overload: 'static member Control.Zero.Zero :  ^t * Internals.Default3 ->  ^t when  ^t : (static member get_Empty : ->  ^t)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a'
.
Possible overload: 'static member Control.Zero.Zero :  ^t * Internals.Default2 ->  ^t when (Control.FromInt32 or  ^t) : (static member FromInt32 :  ^t * Control.FromInt32 -> int32 ->  ^t)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a'
.
Possible overload: 'static member Control.Zero.Zero :  ^t * Internals.Default2 -> ('a1 -> 'a1) when  ^t : null and  ^t : struct'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a'
.
Possible overload: 'static member Control.Zero.Zero :  ^t * Internals.Default1 ->  ^t when  ^t : (static member get_Zero : ->  ^t)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a'
.
Possible overload: 'static member Control.Zero.Zero :  ^t * Internals.Default1 -> ('a1 -> 'a1) when  ^t : null and  ^t : struct'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a'
.
Possible overload: 'static member Control.Zero.Zero : System.TimeSpan * Control.Zero -> System.TimeSpan'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.TimeSpan'
.
Possible overload: 'static member Control.Zero.Zero : Internals.DmStruct * Control.Zero -> Internals.DmStruct'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'Internals.DmStruct'
.
Possible overload: 'static member Control.Zero.Zero : 'a list * Control.Zero -> 'a list'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a list'
.
Possible overload: 'static member Control.Zero.Zero : 'a option * Control.Zero -> 'a option'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a option'
.
Possible overload: 'static member Control.Zero.Zero : 'a array * Control.Zero -> 'a array'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a array'
.
Possible overload: 'static member Control.Zero.Zero : string * Control.Zero -> string'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'string'
.
Possible overload: 'static member Control.Zero.Zero : System.Text.StringBuilder * Control.Zero -> System.Text.StringBuilder'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Text.StringBuilder'
.
Possible overload: 'static member Control.Zero.Zero : unit * Control.Zero -> unit'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'unit'
.
Possible overload: 'static member Control.Zero.Zero : Set<'a> * Control.Zero -> Set<'a> when 'a : comparison'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'Set<'a>'
.
Possible overload: 'static member Control.Zero.Zero : Map<'a,'b> * Control.Zero -> Map<'a,'b> when 'a : comparison'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'Map<'a,'b>'
.
Possible overload: 'static member Control.Zero.Zero : t: ^t * Control.Zero ->  ^t when  ^t : (member get_Item1 :  ^t ->  ^t1) and  ^t : (member get_Item2 :  ^t ->  ^t2) and  ^t : (member get_Item3 :  ^t ->  ^t3) and  ^t : (member get_Item4 :  ^t ->  ^t4) and  ^t : (member get_Item5 :  ^t ->  ^t5) and  ^t : (member get_Item6 :  ^t ->  ^t6) and  ^t : (member get_Item7 :  ^t ->  ^t7) and  ^t : (member get_Rest :  ^t ->  ^tr) and (Control.Zero or  ^t1) : (static member Zero :  ^t1 * Control.Zero ->  ^t1) and (Control.Zero or  ^t2) : (static member Zero :  ^t2 * Control.Zero ->  ^t2) and (Control.Zero or  ^t3) : (static member Zero :  ^t3 * Control.Zero ->  ^t3) and (Control.Zero or  ^t4) : (static member Zero :  ^t4 * Control.Zero ->  ^t4) and (Control.Zero or  ^t5) : (static member Zero :  ^t5 * Control.Zero ->  ^t5) and (Control.Zero or  ^t6) : (static member Zero :  ^t6 * Control.Zero ->  ^t6) and (Control.Zero or  ^t7) : (static member Zero :  ^t7 * Control.Zero ->  ^t7) and (Control.Zero or  ^tr) : (static member Zero :  ^tr * Control.Zero ->  ^tr)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a'
.
Possible overload: 'static member Control.Zero.Zero : System.Tuple< ^a> * Control.Zero -> System.Tuple< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Tuple<'a>'
.
Possible overload: 'static member Control.Zero.Zero : Internals.Id<'a> * Control.Zero -> Internals.Id< ^a1> when (Control.Zero or  ^a1) : (static member Zero :  ^a1 * Control.Zero ->  ^a1)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'Internals.Id<'a>'
.
Possible overload: 'static member Control.Zero.Zero : ( ^a *  ^b) * Control.Zero ->  ^a *  ^b when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a * 'b'
.
Possible overload: 'static member Control.Zero.Zero : ( ^a *  ^b *  ^c) * Control.Zero ->  ^a *  ^b *  ^c when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a * 'b * 'c'
.
Possible overload: 'static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d) * Control.Zero ->  ^a *  ^b *  ^c *  ^d when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a * 'b * 'c * 'd'
.
Possible overload: 'static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d *  ^e) * Control.Zero ->  ^a *  ^b *  ^c *  ^d *  ^e when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d) and (Control.Zero or  ^e) : (static member Zero :  ^e * Control.Zero ->  ^e)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a * 'b * 'c * 'd * 'e'
.
Possible overload: 'static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d *  ^e *  ^f) * Control.Zero ->  ^a *  ^b *  ^c *  ^d *  ^e *  ^f when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d) and (Control.Zero or  ^e) : (static member Zero :  ^e * Control.Zero ->  ^e) and (Control.Zero or  ^f) : (static member Zero :  ^f * Control.Zero ->  ^f)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a * 'b * 'c * 'd * 'e * 'f'
.
Possible overload: 'static member Control.Zero.Zero : ( ^a *  ^b *  ^c *  ^d *  ^e *  ^f *  ^g) * Control.Zero ->  ^a *  ^b *  ^c *  ^d *  ^e *  ^f *  ^g when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a) and (Control.Zero or  ^b) : (static member Zero :  ^b * Control.Zero ->  ^b) and (Control.Zero or  ^c) : (static member Zero :  ^c * Control.Zero ->  ^c) and (Control.Zero or  ^d) : (static member Zero :  ^d * Control.Zero ->  ^d) and (Control.Zero or  ^e) : (static member Zero :  ^e * Control.Zero ->  ^e) and (Control.Zero or  ^f) : (static member Zero :  ^f * Control.Zero ->  ^f) and (Control.Zero or  ^g) : (static member Zero :  ^g * Control.Zero ->  ^g)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a * 'b * 'c * 'd * 'e * 'f * 'g'
.
Possible overload: 'static member Control.Zero.Zero : System.Threading.Tasks.Task< ^a> * Control.Zero -> System.Threading.Tasks.Task< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Threading.Tasks.Task<'a>'
.
Possible overload: 'static member Control.Zero.Zero : ('T ->  ^Monoid) * Control.Zero -> ('T ->  ^Monoid) when (Control.Zero or  ^Monoid) : (static member Zero :  ^Monoid * Control.Zero ->  ^Monoid)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    ''a -> 'b'
.
Possible overload: 'static member Control.Zero.Zero : Async< ^a> * Control.Zero -> Async< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'Async<'a>'
.
Possible overload: 'static member Control.Zero.Zero : Quotations.Expr< ^a> * Control.Zero -> Quotations.Expr< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'Quotations.Expr<'a>'
.
Possible overload: 'static member Control.Zero.Zero : System.Lazy< ^a> * Control.Zero -> Lazy< ^a> when (Control.Zero or  ^a) : (static member Zero :  ^a * Control.Zero ->  ^a)'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Lazy<'a>'
.
Possible overload: 'static member Control.Zero.Zero : System.Collections.Generic.Dictionary<'a,'b> * Control.Zero -> System.Collections.Generic.Dictionary<'a,'b> when 'a : equality'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Collections.Generic.Dictionary<'a,'b>'
.
Possible overload: 'static member Control.Zero.Zero : ResizeArray<'a> * Control.Zero -> ResizeArray<'a>'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'ResizeArray<'a>'
.
Possible overload: 'static member Control.Zero.Zero : seq<'a> * Control.Zero -> seq<'a>'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'seq<'a>'
.
Possible overload: 'static member Control.Zero.Zero : System.Collections.Generic.IEnumerator<'a> * Control.Zero -> System.Collections.Generic.IEnumerator<'a>'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Collections.Generic.IEnumerator<'a>'
.
Possible overload: 'static member Control.Zero.Zero : System.Collections.Generic.IDictionary<'a,'b> * Control.Zero -> System.Collections.Generic.IDictionary<'a,'b> when 'a : equality'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Collections.Generic.IDictionary<'a,'b>'
.
Possible overload: 'static member Control.Zero.Zero : System.Collections.Generic.IReadOnlyDictionary<'a,'b> * Control.Zero -> System.Collections.Generic.IReadOnlyDictionary<'a,'b> when 'a : equality'. Type constraint mismatch. The type
    'MonoidSample'
is not compatible with type
    'System.Collections.Generic.IReadOnlyDictionary<'a,'b>'
```

Note that for operators or generic properties, right now the compiler doesn't show the types given when showing the overload list.

I'm going to work on updating this PR to show the relevant information from `TraitConstraintInfo` when this is the case, after which I think the new error message will be better (right now it doesn't mention `MonoidSample`, like it would do for method calls).

smoothdeveloper commented:
I want to make a case about "the type A is not compatible with type B" being helpful, it is helpful but it is not as helpful as it should, because it doesn't point to a specific argument, and the overall message leaves user too much room to thinker instead of giving the precise location.

Here is a case:

```fsharp
open System.Drawing
let g : Graphics = null
let p : Pen = null
let x = 1.f
let y = 1.f
let w = 1.f
let h = 1

g.DrawRectangle(p, x, y, w, h)
```

currently giving 
```
  g.DrawRectangle(p, x, y, w, h );;
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

stdin(11,1): error FS0041: No overloads match for method 'DrawRectangle'. The available overloads are shown below.
Possible overload: 'Graphics.DrawRectangle(pen: Pen, x: float32, y: float32, width: float32, height: float32) : unit'. Type constraint mismatch. The type
    'int'
is not compatible with type
    'float32'
.
Possible overload: 'Graphics.DrawRectangle(pen: Pen, x: int, y: int, width: int, height: int) : unit'. Type constraint mismatch. The type
    'float32'
is not compatible with type
    'int'
.
```

User can't do much, beside scanning the type of all the parameters anyways, and this is only with two overloads, while in this PR:

```
  g.DrawRectangle(p, x, y, w, h );;
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

stdin(11,1): error FS0041: No overloads match for method 'DrawRectangle'.

Known types of arguments: Pen * float32 * float32 * float32 * int

Available overloads:
 - Graphics.DrawRectangle(pen: Pen, x: float32, y: float32, width: float32, height: float32) : unit
 - Graphics.DrawRectangle(pen: Pen, x: int, y: int, width: int, height: int) : unit
```

it is much easier to see that my last argument has wrong type.

If we had to keep the verbose / per argument mismatch error, it would be if we could precisely point at the mistmatched arguments in the quoted expression, and after initial reporting of available overloads.

Also if we could make it clearer to the user which is which among the expected and given (the phrasing is not clear enough beside pointing incompatibility, it doesn't make clear which is on the signature and which is in the calling side)

```
Possible overload: 'Graphics.DrawRectangle(pen: Pen, x: int, y: int, width: int, height: int) : unit'. Type constraint mismatch. The type
    'float32'
is not compatible with type
    'int'
.
```

smoothdeveloper commented:
This is green, with @dsyme comments addressed (you can see the changes in NicePrint.fs), and some additional improvement ensuring the changes don't remove any info which was originally present, add some more in complex cases where the type checker is yielding a `TraitConstraintInfo`, and actually improve, in general cases, the usefulness to identify which argument is causing mismatch with no second guessing for the user.

Previously it was at best letting the user find it out if the arguments all had different types, or at worse, ambiguous and leaving the user to trial and error.

the simple example for this is the following:

```fsharp
open System.Drawing
let g : Graphics = null
let p : Pen = null
let x = 1.f
let y = 1.f
let w = 1.f
let h = 1

g.DrawRectangle(p, x, y, w, h)
```

originally giving:

>error FS0041: No overloads match for method 'DrawRectangle'. The available overloads are shown below.
Possible overload: 'Graphics.DrawRectangle(pen: Pen, x: float32, y: float32, width: float32, height: float32) : unit'. Type constraint mismatch. The type
    'int'
is not compatible with type
    'float32'
.
Possible overload: 'Graphics.DrawRectangle(pen: Pen, x: int, y: int, width: int, height: int) : unit'. Type constraint mismatch. The type
    'float32'
is not compatible with type
    'int'
.


now turned into

>typecheck error FS0041: No overloads match for method 'DrawRectangle'.
>
>Known types of arguments: Pen * float32 * float32 * float32 * int
>
>Available overloads:
> - Graphics.DrawRectangle(pen: Pen, x: float32, y: float32, width: float32, height: float32) : unit // Argument 'height' doesn't match
> - Graphics.DrawRectangle(pen: Pen, x: int, y: int, width: int, height: int) : unit // Argument 'x' doesn't match

revealing that either `height` or `x` for the respective overloads which are causing the mismatch.

If the arguments are not named, then the one based index is used.

More information is added with regard to the return type and generic type arguments when they are known to be failing the type checking, a complex sample is https://github.com/dotnet/fsharp/blob/9263726bf484df77d2f06ca01aa252186bd195f5/tests/fsharp/typecheck/overloads/neg_known_return_type_and_known_type_arguments.fsx where the call to `zero` fails and will indicate the type parameters to the inline `zero` function defined as

```fsharp
let inline zero< ^Monoid when (Zero or ^Monoid) : (static member Zero : ^Monoid * Zero -> ^Monoid) > : ^Monoid = Zero.Invoke ()
```

The message will now contain this at the top before listing the overloads:

> Known return type: MonoidSample
>
> Known type parameters: < MonoidSample , Zero >

Before, it would only give `MonoidSample` in the type subsumption error, repeated for each overload tried.

This is useful when using SRTP patterns in the same fashion as done in FSharpPlus library.

Please let me know if I missed something or think I need to adjust any of the changes.

cartermp commented:
@dsyme could you look at the guts of this change?

dsyme commented on TESTGUIDE.md:
```diff
@@ -4,11 +4,17 @@
-    build.cmd coreclr test
-    build.cmd vs test
-    build.cmd all test
+    ./build -testAll -c Release
```
This is `.\build` on windows....

smoothdeveloper commented on src/fsharp/NicePrint.fs:
```diff
@@ -1107,6 +1134,46 @@ module private PrintTypes =
+        let argInfos,retTy,genParamTys, cxs =
+          let typesWithDiscrimants =
+            [
+              yield 0uy,retTy 
```
I've updated to int, byte was smallest thing I could think of without defining a type that would be only used locally anyways.

I've added a comment to the new `PrettifyDiscriminantAndTypePairs` explaining the idiom of passing discriminant value with each type to be able to sort them out after the process and clarified what those 0,1,2 means at the place they are used.

smoothdeveloper commented on src/fsharp/FSComp.txt:
```diff
@@ -1,6 +1,6 @@
-# use a completely new error number and add new messages at the end of the file
-# -----------------------------------------------------------------------------
+# -------------------------------------------------------------------------------
+# use a completely new error number and keep messages in their surrounding groups
```
one reason for this is to reduce merge conflicts, when we append at the end, it always conflict, when kept in suroundings, it is unlikely to conflict.

smoothdeveloper commented on tests/fsharp/typecheck/overloads/neg_System.Threading.Tasks.Task.Run.OverloadList.bsl:
```diff
@@ -0,0 +1,48 @@
+
+Candidates:
+ - Task.Run<'TResult>(function: Func<'TResult>) : Task<'TResult>
+ - Task.Run<'TResult>(function: Func<Task<'TResult>>) : Task<'TResult>
```
Note that this error message highlights something where I find the compiler not trying hard enough, here `Func<Task<int>>` is theoritically matching `Func<Task<'TResult>>` overload, in the older compiler, it shows:

> A unique overload for method 'Run' could not be determined based on type information prior to this program point. A type annotation may be needed. Candidates: Task.Run<'TResult>(function: Func<'TResult>) : Task<'TResult>, Task.Run<'TResult>(function: Func<Task<'TResult>>) : Task<'TResult>

Where the user lacked the `Func<Task<int>>` being the known type of argument and had no type subsumption error either.

I feel the ` A type annotation may be needed.` message should turn into more specifically pointing at requesting the generic type argument in that case.

I have not investigated but felt it would be for, worst case, another round of enhancements on that message in separate PR, or best case, adding a step to overload resolution to pick that overload.

@gusty mentionned to me having the spec of current behaviour rewritten more formally, then adding such cases would be the approach.

dsyme replied:
Yes, makes sense as an area for subsequent improvement after this PR

smoothdeveloper commented on tests/fsharp/conformance/lexicalanalysis/E_LessThanDotOpenParen001.bsl:
```diff
@@ -0,0 +1,48 @@
+
+E_LessThanDotOpenParen001.fsx(13,71,13,75): typecheck error FS0193: A type parameter is missing a constraint 'when ( ^T or  ^?8151) : (static member ( + ) :  ^T *  ^?8151 ->  ^?8152)'
```
Seeing this test failing for first time on CI while it passes locally.

> (static member ( + ) :  ^T *  ^?8151 ->  ^?8152)'

I will try again in a release build if that somehow changes the result and is consistent with CI, it used to pass so I'm checking if rebasing on master has any effect.

Originally, this test was not asserting on the whole output and was in fsharpqa, asserting only on a subset of the errors.

It is related to this tracked issue: https://github.com/dotnet/fsharp/issues/6725 and the test will keep showing jumping values with changes done in the compiler.

The error in question is unrelated to the changes in this PR.

smoothdeveloper commented on tests/fsharp/update.base.line.with.actuals.fsx:
```diff
@@ -1,27 +0,0 @@
-open System.IO
```
there was a pre-existing script being more accurate, updated the readme.md as well.

smoothdeveloper commented:
Green again with the comments addressed (unless I overlooked something), I hope it can be vetted before more tests with overload error comes in master 🙂.

smoothdeveloper commented on src/fsharp/MethodCalls.fs:
```diff
@@ -605,7 +635,10 @@ let CoerceFromFSharpFuncToDelegate g amap infoReader ad callerArgTy m callerArgE
 // Handle adhoc argument conversions
 let AdjustCallerArgExprForCoercions (g: TcGlobals) amap infoReader ad isOutArg calledArgTy (reflArgInfo: ReflectedArgInfo) callerArgTy m callerArgExpr = 
-
+  // todo after rebase:
```
Recollecting the full context here: all places calling `AdjustCallerArgExprForCoercions` happen to be just deconstructing `CallerArgs` instances into `isOutArg`, `callerArgTy`, `m` and `callerArgExpr` arguments here.

It felt to me like a potential for refactor (remove the individual parameters and pass the `CallerArgs` record instead) but I wasn't 100% clear about it because some call hardcode `isOutArg` to false.

I had the chance to do a review with @TIHan about this idea, and he pointed that `isOutArg` is guaranted to match at those call sites (only happening in `TcSetterArgExpr`) because this is guarded by `error` if the value is `true`.

I think it is better to do that in a separate PR as it is a pure refactor not related to the error message.

smoothdeveloper commented on src/fsharp/ConstraintSolver.fs:
```diff
@@ -2227,18 +2244,21 @@ and CanMemberSigsMatchUpToCheck
 // "ty2 casts to ty1"
 // "a value of type ty2 can be used where a value of type ty1 is expected"
-and private SolveTypeSubsumesTypeWithReport (csenv: ConstraintSolverEnv) ndeep m trace cxsln ty1 ty2 =
+and private SolveTypeSubsumesTypeWithWrappedContextualReport (csenv: ConstraintSolverEnv) ndeep m trace cxsln ty1 ty2 wrapper =
```
this is called many times, unsure how to make if there is a way to make the change a `zero cost abstraction` when `id` is used as `wrapper`, I wasn't eager to duplicate the logic which is sensitive.

smoothdeveloper commented on src/fsharp/FSComp.txt:
```diff
@@ -361,9 +361,16 @@ csCtorHasNoArgumentOrReturnProperty,"The object constructor '%s' has no argument
 csNoOverloadsFound,"No overloads match for method '%s'."
+csNoOverloadsFoundArgumentsPrefixSingular,"Known type of argument: %s"
+csNoOverloadsFoundArgumentsPrefixPlural,"Known types of arguments: %s"
+csNoOverloadsFoundTypeParametersPrefixSingular,"Known type parameter: %s"
```
I'm not clear if the terminology is right, this is used to display info from `TraitConstraintInfo.ArgTys`

smoothdeveloper replied:
let's address this in separate PR if needed

dsyme commented on src/fsharp/TypeChecker.fs:
```diff
@@ -10476,7 +10477,7 @@ and TcMethodArg cenv env (lambdaPropagationInfo, tpenv) (lambdaPropagationInfoFo
                      yield info |]
 
-    CallerArg(argTy, mArg, isOpt, e'), (lambdaPropagationInfo, tpenv)
+    { Type = argTy; Range = mArg; IsOptional = isOpt; Expr = e' }, (lambdaPropagationInfo, tpenv)
```
Could you undo this change to CallerArg pleae?  It's cosmetic and isn't essential to your improvement, thanks!!

smoothdeveloper replied:
@dsyme, done!

I want to bring attention to few points though:

* I initially did the change because in the debugger variable inspector it shows everything twice
* the fields were unammed (they now are) so you'd see Item1 etc, not great when you aren't familiar with the code and start looking into it
* you can't do an exhaustive find usages on fields, with record you always can
* when properties are defined, two ways to reach the same data in code
* minor, but field access vs property access 

What is the right approach to bring such improvement (affects documentation, debuggability, approachability, searchability) in the compiler codebase?

dsyme commented:
@smoothdeveloper Thank you for minimizing the diff!

smoothdeveloper commented:
@dsyme, no problem, I just rebased against master to fix the CompilerAssert.fs conflict, and now green again.

@cartermp, @dsyme was asking about the xlf, do we want them reverted? (and can they ultimately move out of the repo / git ignored if so)

Can the terminology of messages showing the data pulled from `TraitConstraintInfo` be assessed as I think there was no exchange on this and it is new information coming with this PR.

Thanks a lot.

smoothdeveloper commented:
I've removed the xlf from the diff, they should be added to .gitignore, and in case external contributors want to contribute translations, a guideline written to enable it; as of now, xlf seems like an internal MSFT thing with no impact on contributions to this repository beside generating more questions / diff / merge conflicts.

@dsyme, I've fixed the recent conflict and it is (allegedly soon) green again.

@cartermp said on slack it were now dependent on your aproval and I feel our degree of confidence about the changes have passed the bar, also the change is as minimal as we'd want to not disrupt other branches were work is happening.

I'll definitely be around on simple ping for any adjustment needed on top of this work if needed, in separate PRs, but maintaing this long lived branch through merged changes I don't have good grasp on is becoming harder and increasing the overhead of impact versus cumulated time spent by everyone involved.

Once merged, Ionide and Rider users will also benefit from the changes once they land in FCS, leading to potentially more input / feedback from end users before this ships in Visual Studio / dotnet SDK.

The efforts I spend on rebasing without glitching / losing anything, and the lack of better messages in ongoing work dealing with method overload are adding extra incentives to get it merged, let's do it 🙂.

smoothdeveloper commented:
It is really green.

dsyme commented on src/fsharp/MethodCalls.fs:
```diff
@@ -977,7 +1002,10 @@ let CoerceFromFSharpFuncToDelegate g amap infoReader ad callerArgTy m callerArgE
 // Handle adhoc argument conversions
 let AdjustCallerArgExprForCoercions (g: TcGlobals) amap infoReader ad isOutArg calledArgTy (reflArgInfo: ReflectedArgInfo) callerArgTy m callerArgExpr = 
-
+   // todo after rebase:
```
Please remove this comment?

smoothdeveloper commented:
Comment removed, thanks @dsyme!

cartermp commented:
@KevinRansom wanna take a look?

cartermp commented:
Yeet

KevinRansom commented:
Is yeet good or bad ?  I thought it was throwing things of a cliff?

baronfel commented:
sometimes you have to yeet things to see them fly
